### PR TITLE
Download statistics in submission page

### PR DESCRIPTION
**Pull request by huangjosh**

### CODE REVIEW

File: app/services/course/assessment/submission/statistics_download_service.rb [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+class Course::Assessment::Submission::StatisticsDownloadService
+  class << self
+    # Downloads the submissions statistics and zip it.
+    #
+    # @param [CourseUser] course_user The course user downloading the statistics.
+    # @param [Course::Assessment] assessment The assessments to download statistics from.
+    # @param [String|nil] students The subset of students whose submissions' statistics to download.
+    # Accepted values: 'my', 'phantom'
+    # @return [String] The path to the zip file.
+    def download_and_zip(course_user, assessment, students)
```
Comment (weiqingtoh) [inline app/services/course/assessment/submission/statistics_download_service.rb]: Can I suggest we handle the params 'my' and 'phantom' at the controller? 

That would simplify this service to just accept 1 parameter, which are the collection of submissions to be zipped and downloaded.

File: app/helpers/application_formatters_helper.rb [hunk 2]
```diff
@@ -94,9 +94,9 @@ def display_topic(topic, options = {})
   #        :date_only_long => "December 04, 2007"
   #        :date_only_short => "04 Dec"
   # @return [String] the formatted datetime string
-  def format_datetime(date, format = :long)
-    user_zone = (current_user ? current_user.time_zone : nil) ||
-                Application.config.x.default_user_time_zone
+  def format_datetime(date, format = :long, user = nil)
```
Comment (allenwq) [inline app/helpers/application_formatters_helper.rb]: Prefer to use keyword arguments if arg list is too long. `format_datetime(date, format: :long, user:)`

File: app/models/course/assessment/submission.rb [hunk 3]
```diff
@@ -237,4 +278,49 @@ def phantom_submission_email_enabled?
     Course::Settings::AssessmentsComponent.
       email_enabled?(assessment.tab.category, :new_phantom_submission)
   end
+
+  def csv_empty
```
Comment (allenwq) [inline app/models/course/assessment/submission.rb]: Probably should shift all these methods to service, since it will be only used for download service and we prefer to only keep the methods that are used globally in the model.

File: app/helpers/application_formatters_helper.rb (hunk 2 above)
Comment (huangjosh) [reply app/helpers/application_formatters_helper.rb]: This method has already been used in too many places, we need to change a lot, if we make the ```format``` as a key word argument. I would prefer only make the user as a keyword argument?

File: app/models/course/assessment/submission.rb (hunk 3 above)
Comment (huangjosh) [reply app/models/course/assessment/submission.rb]: Ok, I think I just shift the ```download_statistics``` and ```download_statistics_header``` to service as well?

File: app/services/course/assessment/submission/statistics_download_service.rb (hunk 1 above)
Comment (huangjosh) [reply app/services/course/assessment/submission/statistics_download_service.rb]: OK, no problem, will change that.

File: app/helpers/application_formatters_helper.rb (hunk 2 above)
Comment (allenwq) [reply app/helpers/application_formatters_helper.rb]: OK, sure

File: app/models/course/assessment/submission.rb (hunk 3 above)
Comment (allenwq) [reply app/models/course/assessment/submission.rb]: 👍 yep, thanks! and also include the helper only in the service

**Comment by huangjosh** [comment]
@weiqingtoh @allenwq Fixed, and updated.

File: app/models/course/assessment/submission.rb [hunk 4]
```diff
@@ -169,6 +169,20 @@ def grade
     latest_answers.map { |a| a.grade || 0 }.sum
   end
 
+  def last_graded_at
```
Comment (allenwq) [inline app/models/course/assessment/submission.rb]: there's a calculated attribute of `last_graded_at`, I think this method is unnecessary.

File: app/models/course/assessment/submission.rb [hunk 5]
```diff
@@ -169,6 +169,20 @@ def grade
...
+    current_graded_at = nil
+    answers.each do |answer|
+      if answer.graded_at && (current_graded_at.nil? || current_graded_at < answer.graded_at)
+        current_graded_at = answer.graded_at
+      end
+    end
+    current_graded_at
+  end
+
+  def maximum_grade
```
Comment (allenwq) [inline app/models/course/assessment/submission.rb]: assessment has `maximum_grade`, I think you might want `submission.assessment.maximum_grade`? And if you does this, also need to eager load assessments.

File: app/services/course/assessment/submission/statistics_download_service.rb [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+      format_datetime(submission.last_graded_at, :long, user: @course_user.user)
+    else
+      csv_empty
+    end
+  end
+
+  def csv_grading_time(submission)
+    if submission.last_graded_at && submission.submitted_at
+      ActionController::Base.
+        helpers.distance_of_time_in_words(submission.last_graded_at - submission.submitted_at)
```
Comment (allenwq) [inline app/services/course/assessment/submission/statistics_download_service.rb]: I am not sure why this field is in human readable format, I suspect that is hard to parse in CSV.  Are there anything that I am missing?

File: app/services/course/assessment/submission/statistics_download_service.rb [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    statistics_path = File.join(statistic_dir, 'statistics.csv')
+    CSV.open(statistics_path, 'w') do |csv|
+      @submissions.each_with_index do |submission, idx|
+        download_statistics_header csv if idx == 0
+        download_statistics csv, submission
+      end
+    end
+  end
+
+  def create_folder(parent, folder_name)
```
Comment (allenwq) [inline app/services/course/assessment/submission/statistics_download_service.rb]: Any reasons that you want to create folders and zip the CSV file?

File: app/services/course/assessment/submission/statistics_download_service.rb [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        service.send(:download_to_base_dir)
+      end
+      service.send(:zip_base_dir)
+    end
+  end
+
+  private
+
+  def initialize(course_user, submissions)
+    @course_user = course_user
```
Comment (allenwq) [inline app/services/course/assessment/submission/statistics_download_service.rb]: Should this be `@current_user` instead of `@course_user`? Don't think we need any info from course_user

File: client/app/bundles/course/assessment/submission/pages/SubmissionsIndex/SubmissionsTable.jsx [hunk 9]
```diff
@@ -1,10 +1,13 @@
 // eslint-disable-next-line import/extensions, import/no-extraneous-dependencies, import/no-unresolved
 import React from 'react';
 import PropTypes from 'prop-types';
-import { FormattedMessage } from 'react-intl';
+import { FormattedMessage, injectIntl, intlShape } from 'react-intl';
```
[the remark this answers is not in the document]
Comment (huangjosh) [reply client/app/bundles/course/assessment/submission/pages/SubmissionsIndex/SubmissionsTable.jsx]: But ```FormattedMessage``` is a react node, I think ```primaryText``` can only be a string?

File: app/services/course/assessment/submission/statistics_download_service.rb (hunk 8 above)
Comment (huangjosh) [reply app/services/course/assessment/submission/statistics_download_service.rb]: Will change to current user instead.

File: app/services/course/assessment/submission/statistics_download_service.rb (hunk 6 above)
Comment (huangjosh) [reply app/services/course/assessment/submission/statistics_download_service.rb]: Let's use minutes instead?

Comment (allenwq) [reply app/services/course/assessment/submission/statistics_download_service.rb]: how about `HH:MM:SS`? any specific requirement for this?

Comment (huangjosh) [reply app/services/course/assessment/submission/statistics_download_service.rb]: Nope. We have another column called grading date/time. This column is for the time period for grading.

Comment (allenwq) [reply app/services/course/assessment/submission/statistics_download_service.rb]: How about https://en.wikipedia.org/wiki/ISO_8601

Duration will be sth like`3H2M5S`

File: app/models/course/assessment/submission.rb (hunk 4 above)
Comment (huangjosh) [reply app/models/course/assessment/submission.rb]: I did not find this attribute?

Comment (allenwq) [reply app/models/course/assessment/submission.rb]: Oops, it's => https://github.com/Coursemology/coursemology2/blob/master/app/models/course/assessment/submission.rb#L77

It's a calculated attribute, one of the gem that we are using: https://github.com/aha-app/calculated_attributes

File: client/app/bundles/course/assessment/submission/pages/SubmissionsIndex/SubmissionsTable.jsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (allenwq) [reply client/app/bundles/course/assessment/submission/pages/SubmissionsIndex/SubmissionsTable.jsx]: Hmm OK, I didn't know that it only supports string. If that's the case then it's fine.

File: app/models/course/assessment/submission.rb (hunk 4 above)
Comment (huangjosh) [reply app/models/course/assessment/submission.rb]: Got it.:P

File: app/services/course/assessment/submission/statistics_download_service.rb [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    @submissions = submissions
```
[the remark this answers is not in the document]
Comment (huangjosh) [reply app/services/course/assessment/submission/statistics_download_service.rb]: According to weiqingtoh's comment, I moved query into the controller. The problem of ```ApplicationJob``` is it cannot receive Association as the input parameter. Hence, all the eager load in the controller won't help in the service. 
Hence, I think we need to move the query back to the service?

File: client/app/bundles/course/assessment/submission/pages/SubmissionsIndex/SubmissionsTable.jsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (huangjosh) [reply client/app/bundles/course/assessment/submission/pages/SubmissionsIndex/SubmissionsTable.jsx]: My bad. I double checked, it supports node. I will change this one.

File: app/models/course/assessment/submission.rb (hunk 5 above)
Comment (huangjosh) [reply app/models/course/assessment/submission.rb]: We will pass in assessment instead of using assessment from the submission.

File: app/services/course/assessment/submission/statistics_download_service.rb (hunk 6 above)
Comment (huangjosh) [reply app/services/course/assessment/submission/statistics_download_service.rb]: OK.

File: app/services/course/assessment/submission/statistics_download_service.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (huangjosh) [reply app/services/course/assessment/submission/statistics_download_service.rb]: May I ask if ```graded_at``` will benefit from the eager load of answers?  Since ```graded_at``` does not use Associations, and it use plain sql?

File: client/app/bundles/course/assessment/submission/pages/SubmissionsIndex/SubmissionsTable.jsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (allenwq) [reply client/app/bundles/course/assessment/submission/pages/SubmissionsIndex/SubmissionsTable.jsx]: 😄 thanks for checking!

File: app/services/course/assessment/submission/statistics_download_service.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (weiqingtoh) [reply app/services/course/assessment/submission/statistics_download_service.rb]: > According to weiqingtoh's comment, I moved query into the controller. The problem of ```ApplicationJob``` is it cannot receive Association as the input parameter. Hence, all the eager load in the controller won't help in the service. 
Hence, I think we need to move the query back to the service? 

You can just `pluck` the submission_ids, and send that as a param into the service, rather than sending the entire set of submission objects. Then in the service, do the necessary loading eager loading for the service - so when the service is updated next time, the engineer just needs to modify this in the `initialize` method. 

> May I ask if ```graded_at``` will benefit from the eager load of answers?  Since ```graded_at``` does not use Associations, and it use plain sql? 

It depends on the implementation. For our case, `calculated_attributes` gets `graded_at` through a pure SQL injected as a sub-query, so including the association does not provide any benefit (since it is still done in 1 query). 

If for some reason we use a ruby method to calculate `graded_at`, each `submission` would require the `answers` association, and yes it might be better to eager load then. 

---

Hope this clarifies @huangjosh

[the remark this answers is not in the document]
Comment (huangjosh) [reply app/services/course/assessment/submission/statistics_download_service.rb]: Thanks @weiqingtoh . This is very helpful.
I think I will follow you suggestion to improve this PR.

**Comment by huangjosh** [comment]
@weiqingtoh @allenwq Fixed all the issues and updated the PR. :)

File: app/services/course/assessment/submission/statistics_download_service.rb [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+  include ApplicationFormattersHelper
+
+  class << self
+    # Downloads the statistics and zip them.
+    #
+    # @param [User] current_user The current user downloading the statistics.
+    # @param [Course::Assessment] assessment The assessments to download  from.
+    # @param [Array<Integer>] submission_ids The ids of the submissions to download statistics for
+    # @return [String] The path to the zip file.
+    def download(current_user, assessment, submission_ids)
```
Comment (weiqingtoh) [inline app/services/course/assessment/submission/statistics_download_service.rb]: Actually if you wish to make the service even cleaner we can get `@assessment` from the `submission_ids`, and just retrieve the maximum grade. Might be easier when testing the service.

File: app/services/course/assessment/submission/statistics_download_service.rb [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+    @submission_ids = submission_ids
+    @base_dir = Dir.mktmpdir('coursemology-statistics-')
+  end
+
+  def download_to_base_dir
+    submissions = Course::Assessment::Submission.
+                  where(id: @submission_ids).includes(:course_user, :answers)
+    statistics_file_path = File.join(@base_dir, 'statistics.csv')
+    CSV.open(statistics_file_path, 'w') do |csv|
+      submissions.each_with_index do |submission, idx|
```
Comment (weiqingtoh) [inline app/services/course/assessment/submission/statistics_download_service.rb]: Why not just
```ruby
CSV.open(statistics_file_path, 'w') do |csv|
  download_statistics_header csv
  submissions.each do { |submission| download_statistics csv, submission }
end
```

File: app/controllers/course/assessment/submission/submissions_controller.rb [hunk 13]
```diff
@@ -115,6 +117,17 @@ def download_all
     end
   end
 
+  def download_statistics
+    authorize!(:manage, @assessment)
+    submission_ids = @assessment.submissions.by_users(student_ids).pluck(:id)
```
Comment (weiqingtoh) [inline app/controllers/course/assessment/submission/submissions_controller.rb]: How do we want to handle this if `submission_ids` is `nil`?

File: app/services/course/assessment/submission/statistics_download_service.rb [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+class Course::Assessment::Submission::StatisticsDownloadService
+  include ApplicationFormattersHelper
+
+  class << self
+    # Downloads the statistics and zip them.
+    #
+    # @param [User] current_user The current user downloading the statistics.
+    # @param [Course::Assessment] assessment The assessments to download  from.
+    # @param [Array<Integer>] submission_ids The ids of the submissions to download statistics for
+    # @return [String] The path to the zip file.
```
Comment (weiqingtoh) [inline app/services/course/assessment/submission/statistics_download_service.rb]: Should update this to Path to the CSV file instead.

File: app/services/course/assessment/submission/statistics_download_service.rb [hunk 15]
```diff
@@ -0,0 +1,120 @@
+# frozen_string_literal: true
+require 'csv'
+class Course::Assessment::Submission::StatisticsDownloadService
```
Comment (weiqingtoh) [inline app/services/course/assessment/submission/statistics_download_service.rb]: Is it possible to add a test for this service?

**Comment by weiqingtoh** [comment]
Last set of comments @huangjosh, nearly there! 👍

File: app/services/course/assessment/submission/statistics_download_service.rb [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+  private
+
+  def initialize(current_user, assessment, submission_ids)
+    @current_user = current_user
+    @assessment = assessment
+    @submission_ids = submission_ids
+    @base_dir = Dir.mktmpdir('coursemology-statistics-')
+  end
+
+  def download_to_base_dir
```
Comment (allenwq) [inline app/services/course/assessment/submission/statistics_download_service.rb]: 😄 pehaps we can rename this method to `generate_csv_report`? (or some other names which reveal the intent)

File: app/services/course/assessment/submission/statistics_download_service.rb (hunk 11 above)
Comment (huangjosh) [reply app/services/course/assessment/submission/statistics_download_service.rb]: OK, will remove ```assessment```

File: app/services/course/assessment/submission/statistics_download_service.rb (hunk 12 above)
Comment (huangjosh) [reply app/services/course/assessment/submission/statistics_download_service.rb]: Ok, will remove this.

File: app/controllers/course/assessment/submission/submissions_controller.rb (hunk 13 above)
Comment (huangjosh) [reply app/controllers/course/assessment/submission/submissions_controller.rb]: I don't think ```submission_ids``` can be null. It can only be empty array?

File: app/services/course/assessment/submission/statistics_download_service.rb (hunk 14 above)
Comment (huangjosh) [reply app/services/course/assessment/submission/statistics_download_service.rb]: OK, I will fix this.

File: app/services/course/assessment/submission/statistics_download_service.rb (hunk 16 above)
Comment (huangjosh) [reply app/services/course/assessment/submission/statistics_download_service.rb]: OK, will rename the method.

File: app/services/course/assessment/submission/statistics_download_service.rb (hunk 15 above)
Comment (huangjosh) [reply app/services/course/assessment/submission/statistics_download_service.rb]: OK, I will add test case as well.

**Comment by huangjosh** [comment]
@weiqingtoh @allenwq Fixed and updated the PR.

File: app/controllers/course/assessment/submission/submissions_controller.rb (hunk 13 above)
Comment (weiqingtoh) [reply app/controllers/course/assessment/submission/submissions_controller.rb]: Yup. Actually my concern is that if it is an empty array should we return some 'error message' (eg. `no submissions are selected`) instead of downloading an empty CSV file?

Comment (huangjosh) [reply app/controllers/course/assessment/submission/submissions_controller.rb]: There is no detail requirement for this. I will fix this. I will throw bad request, if the submission id is empty.

**Comment by huangjosh** [comment]
@weiqingtoh @allenwq Fixed and updated the PR. There is an API change on ```setNotification``` method. It will support both intl message object and string.

File: app/controllers/course/assessment/submission/submissions_controller.rb (hunk 13 above)
Comment (weiqingtoh) [reply app/controllers/course/assessment/submission/submissions_controller.rb]: Ok thanks! 👍

File: client/app/lib/components/IntlNotificationBar.jsx [hunk 17]
```diff
@@ -1,26 +1,29 @@
...
 
 import NotificationBar from 'lib/components/NotificationBar';
 
 export const notificationShape = PropTypes.shape({
-  message: PropTypes.shape({
-    id: PropTypes.string.isRequired,
-  }),
+  message: PropTypes.oneOfType([
+    PropTypes.shape({
+      id: PropTypes.string.isRequired,
```
Comment (allenwq) [inline client/app/lib/components/IntlNotificationBar.jsx]: We already have a NotificationBar component under for normal strings. Should you use that instead? I am ok if you can combine these two components into one.

File: client/package.json [hunk 18]
```diff
@@ -75,6 +75,7 @@
     "jquery": "2.2.4",
     "jquery-ui": "^1.12.1",
     "json-loader": "^0.5.4",
+    "lodash": "^4.17.4",
```
Comment (allenwq) [inline client/package.json]: We have discussed this before, we'd prefer to not use lodash if only a few places benefits from it. There're costs in extra bundle size and learning curve.  Would appreciate if you can implement the notification without it.

**Comment by huangjosh** [comment]
@weiqingtoh @allenwq I remove the ```lodash``` dependence. Added and new commit to combine function of ```IntlNotificationBar``` into ```NotificationBar```. Could you help to take a look.

**Comment by allenwq** [comment]
👍  Thanks @huangjosh ! 

Build failed, LGTM otherwise.

**Comment by huangjosh** [comment]
@allenwq My bad. I have just fixed, it.:)
